### PR TITLE
Support RNNLM shallow fusion in modified beam search

### PR DESCRIPTION
**Pull request by marcoyang1998**
This PR adds RNNLM shallow fusion during `modified_beam_search` to improve WERs.

Here are the WERs after shallow fusion:

`lstm_transducer_stateless2`
| Decoding method | test clean | test-other |
| --- |------------- | ------------- |
| modified_beam_search |2.73 | 7.15  |
| modified_beam_search  + RNNLM shallow fusion | 2.42 |  6.46  |

`pruned_transducer_stateless5`
| Decoding method | test clean | test-other |
| --- |------------- | ------------- |
| modified_beam_search | 2.47 | 5.71  |
| modified_beam_search  + RNNLM shallow fusion | 2.27 |  5.24  |

It would be very nice if someone can propagate these changes to other recipes!

### CODE REVIEW

File: egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py [hunk 1]
```diff
@@ -1851,3 +1853,224 @@ def modified_beam_search_ngram_rescoring(
...
+    sos_token = torch.tensor([[sos_id]]).to(torch.int64).to(device)
+    init_score, init_states = rnnlm.score_token(sos_token)
+
+    B = [HypothesisList() for _ in range(N)]
+    for i in range(N):
+        B[i].add(
+            Hypothesis(
+                ys=[blank_id] * context_size,
+                log_prob=torch.zeros(1, dtype=torch.float32, device=device),
+                state=init_states,
```
Comment (ezerhouni) [inline egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py:1924]: Does Hypothesis has ```state``` as class variable ? ?

File: egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py [hunk 2]
```diff
@@ -1851,3 +1853,224 @@ def modified_beam_search_ngram_rescoring(
...
+                hyp_log_prob = topk_log_probs[k]  # get score of current hyp
+                new_token = topk_token_indexes[k]
+                if new_token not in (blank_id, unk_id):
+
+                    ys.append(new_token)
+                    hyp_log_prob += (
+                        lm_score[new_token] * lm_scale
+                    )  # add the lm score
+
+                    lm_score = scores[count]
```
Comment (ezerhouni) [inline egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py:2055]: If ```token_list``` is empty, ```scores``` and ```lm_states``` are not defined

File: egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py [hunk 3]
```diff
@@ -1851,3 +1853,224 @@ def modified_beam_search_ngram_rescoring(
...
+
+                ys = hyp.ys[:]
+
+                lm_score = hyp.lm_score
+                state = hyp.state
+
+                hyp_log_prob = topk_log_probs[k]  # get score of current hyp
+                new_token = topk_token_indexes[k]
+                if new_token not in (blank_id, unk_id):
+
```
Comment (ezerhouni) [inline egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py:2049]: Could you remove the empty line ?

File: egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py [hunk 4]
```diff
@@ -1851,3 +1853,224 @@ def modified_beam_search_ngram_rescoring(
...
+                    lm_score = scores[count]
+                    state = (
+                        lm_states[0][:, count, :].unsqueeze(1),
+                        lm_states[1][:, count, :].unsqueeze(1),
+                    )
+                    count += 1
+
+                new_hyp = Hypothesis(
+                    ys=ys, log_prob=hyp_log_prob, state=state, lm_score=lm_score
```
Comment (ezerhouni) [inline egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py:2063]: I don't think ```lm_score``` exist in ```Hypothesis```

File: egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py [hunk 5]
```diff
@@ -1851,3 +1853,224 @@ def modified_beam_search_ngram_rescoring(
...
+                lm_score = hyp.lm_score
+                state = hyp.state
+
+                hyp_log_prob = topk_log_probs[k]  # get score of current hyp
+                new_token = topk_token_indexes[k]
+                if new_token not in (blank_id, unk_id):
+
+                    ys.append(new_token)
+                    hyp_log_prob += (
+                        lm_score[new_token] * lm_scale
```
Comment (ezerhouni) [inline egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py:2052]: If I understand correctly, the ```lm_scale``` is for scaling the previous lm_score (not the one from the RNN-LM), is that correct ?

File: egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py [hunk 6]
```diff
@@ -1851,3 +1853,224 @@ def modified_beam_search_ngram_rescoring(
...
+                hyp = A[i][hyp_idx]
+
+                new_token = topk_token_indexes[k]
+                if new_token not in (blank_id, unk_id):
+
+                    assert new_token != 0, new_token
+                    token_list.append([new_token])
+                    hs.append(hyp.state[0])
+                    cs.append(hyp.state[1])
+        # forward RNNLM to get new states and scores
```
Comment (ezerhouni) [inline egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py:2015]: Could you add a empty line for readability ?

**Review by ezerhouni — COMMENTED** [review]
Some comments, overall looks good to me.
In general, I feel like adding couple of comments could be helpful to follow
Thank you @marcoyang1998

**Comment by ezerhouni** [comment]
@marcoyang1998 I can help propagate the change if you want. However, I am not sure if I can compute the WER for the other recipes. If it is fine with you, would you rather merge that PR first, or you want me to add the commit on top of it

File: egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py (hunk 2 above)
Comment (marcoyang1998) [reply egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py:2055]: If `token_list` is empty, this part will also not be reached. But I admit this is a little unclear. Will think of a way of changing it.

**Comment by marcoyang1998** [comment]
@ezerhouni I think we can wait for a little bit. It's not that urgent. What do you think @csukuangfj ?

File: egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py (hunk 4 above)
Comment (marcoyang1998) [reply egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py:2063]: Sry, I deleted them accidentally.

File: egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py [hunk 7]
```diff
@@ -1851,3 +1853,224 @@ def modified_beam_search_ngram_rescoring(
...
+    init_score, init_states = rnnlm.score_token(sos_token)
+
+    B = [HypothesisList() for _ in range(N)]
+    for i in range(N):
+        B[i].add(
+            Hypothesis(
+                ys=[blank_id] * context_size,
+                log_prob=torch.zeros(1, dtype=torch.float32, device=device),
+                state=init_states,
+                lm_score=init_score.reshape(-1),
```
Comment (csukuangfj) [inline egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py:1925]: Also, it does not have `lm_score`.

File: egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py (hunk 2 above)
Comment (ezerhouni) [reply egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py:2055]: Ah yes I see what you mean !

File: egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py (hunk 7 above)
Comment (marcoyang1998) [reply egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py:1925]: Will push the new commits

File: egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py (hunk 5 above)
Comment (marcoyang1998) [reply egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py:2052]: It's not a typo. Each time when we create a new hypothesis, we pass a variable `lm_score` to it, which is the RNNLM score for the next token given the current transcription, i.e `ys`. Therefore, when we encounter a non-blank token, we just need to add `hyp.lm_score * lm_scale` to `hyp.log_prob`

Comment (ezerhouni) [reply egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py:2067]: Ah got it. Another question though, is there a reason why when rescoring with an N-gram, the lm_score is removed :
```
# We only keep AM scores in new_hyp.log_prob
                new_log_prob = (
                    topk_log_probs[k] - hyp.state_cost.lm_score * lm_scale
                )
```

but not in this case ?

Comment (marcoyang1998) [reply egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py:2067]: Well, we also don't know for sure. But I am investigating this.

Also, could you please remind me which text corpora was the RNNLM in this repo (https://huggingface.co/ezerhouni/icefall-librispeech-rnn-lm/tree/main) trained on?

Comment (ezerhouni) [reply egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py:2067]: I trained on librispeech train data : 
https://github.com/k2-fsa/icefall/blob/master/egs/librispeech/ASR/prepare.sh#L283

**Comment by ezerhouni** [comment]
@marcoyang1998 Once your PR is ready, you can merge it and I will add propagate the changes in other places.

**Comment by marcoyang1998** [comment]
Merging this PR now.

File: egs/librispeech/ASR/beam_search.py [hunk 8]
```diff
@@ -0,0 +1,1821 @@
+# Copyright    2021  Xiaomi Corp.        (authors: Fangjun Kuang
```
Comment (csukuangfj) [inline egs/librispeech/ASR/beam_search.py:1]: Shall we put this file in some subfolder?

Comment (marcoyang1998) [reply egs/librispeech/ASR/beam_search.py:1]: Sry, just deleted it.

File: egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py (hunk 5 above)
Comment (marcoyang1998) [reply egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py:2071]: > Ah got it. Another question though, is there a reason why when rescoring with an N-gram, the lm_score is removed :
> 
> ```
> # We only keep AM scores in new_hyp.log_prob
>                 new_log_prob = (
>                     topk_log_probs[k] - hyp.state_cost.lm_score * lm_scale
>                 )
> ```
> 
> but not in this case ?

I found the reason. The n-gram LM returns a cumulative score given the transcription whereas the RNNLM just gives the score for the next token.

**Comment by RuABraun** [comment]
Hi @marcoyang1998 interesting work. could you share the size/config of the rnnlm you used?

**Comment by marcoyang1998** [comment]
Here are the RNNLM configs:

{'max_sent_len': 200, 'sos_id': 1, 'eos_id': 1, 'blank_id': 0, 'lr': 0.001, 'weight_decay': 1e-06, 'best_train_loss': inf, 'best_valid_loss': inf, 'best_train_epoch': -1, 'best_valid_epoch': -1, 'batch_idx_train': 0, 'log_interval': 200, 'reset_interval': 2000, 'valid_interval': 10000, 'env_info': {'k2-version': '1.15.1', 'k2-build-type': 'Release', 'k2-with-cuda': True, 'k2-git-sha1': 'c11c0b70e91d24935514b73d6bffddc8f5a07932', 'k2-git-date': 'Sat Jun 4 06:06:20 2022', 'lhotse-version': '1.3.0.dev+git.4198446.clean', 'torch-version': '1.11.0', 'torch-cuda-available': True, 'torch-cuda-version': '11.3', 'python-version': '3.8', 'icefall-git-branch': 'rnn-lm', 'icefall-git-sha1': '71a9c33-clean', 'icefall-git-date': 'Thu Jun 16 09:17:51 2022', 'icefall-path': '/workspace/package/k2-icefall', 'k2-path': '/opt/conda/lib/python3.8/site-packages/k2-1.15.1.dev20220607+cuda11.3.torch1.11.0-py3.8-linux-x86_64.egg/k2/__init__.py', 'lhotse-path': '/opt/conda/lib/python3.8/site-packages/lhotse/__init__.py',}, 'world_size': 8, 'master_port': 12354, 'tensorboard': True, 'num_epochs': 30, 'start_epoch': 0, 'exp_dir': PosixPath('models/rnn_tied_2048_3'), 'use_fp16': True, 'batch_size': 400, 'use_ddp_launch': False, 'lm_data': 'data/lm_training_bpe_500/sorted_lm_data.pt', 'lm_data_valid': 'data/lm_training_bpe_500/sorted_lm_data-valid.pt', 'vocab_size': 500, 'embedding_dim': 2048, 'hidden_dim': 2048, 'num_layers': 3, 'tie_weights': True, 'seed': 42}

This RNNLM is provided by @ezerhouni. It is uploaded here: https://huggingface.co/ezerhouni/icefall-librispeech-rnn-lm/tree/main
